### PR TITLE
GHC / Haskell packages, wasm cross compilation support

### PR DESCRIPTION
**Pull request by dfordivam**
###### Motivation for this change
These are needed to building GHC and Haskell packages for wasm target, and are being used by WebGHC.

Except the `STRIP_CMD = :`, all other changes are under the `isWasm` target flag, so should not impact any other target's compilation.

These are fairly small changes, so I clubbed them together in one PR. 

###### Things done

- [ ] Tested using sandboxing ([nix.useSandbox](https://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](https://nixos.org/nix/manual/#sec-conf-file) on non-NixOS linux)
- Built on platform(s)
   - [ ] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"`
- [ ] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [ ] Ensured that relevant documentation is up to date
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

### CODE REVIEW

File: pkgs/development/haskell-modules/make-package-set.nix [hunk 1]
```diff
@@ -133,7 +133,7 @@ let
       installPhase = ''
         export HOME="$TMP"
         mkdir -p "$out"
-        cabal2nix --compiler=${self.ghc.haskellCompilerName} --system=${hostPlatform.config} ${sha256Arg} "${src}" ${extraCabal2nixOptions} > "$out/default.nix"
+        cabal2nix --compiler=${self.ghc.haskellCompilerName} ${stdenv.lib.optionalString (!hostPlatform.isWasm) "--system=${hostPlatform.config}"} ${sha256Arg} "${src}" ${extraCabal2nixOptions} > "$out/default.nix"
```
Comment (Ericson2314) [inline pkgs/development/haskell-modules/make-package-set.nix:136]: I suppose this is fine for now, but can you add a "TODO patch Cabal?"

**Review by Ericson2314 — COMMENTED** [review]
Even though WebGHC doesn't support the older GHCs, could we apply these changes to them too? Should be harmless and helps avoid them accidentally drifting out of sync. (And keeps alive the hope that somebody will finally write an abstraction to replace the copy-paste.)

File: pkgs/development/haskell-modules/make-package-set.nix (hunk 1 above)
Comment (dfordivam) [reply pkgs/development/haskell-modules/make-package-set.nix]: Got rid of this, on the latest nixpkgs things work fine without this change.

File: pkgs/development/haskell-modules/generic-builder.nix [hunk 2]
```diff
@@ -207,7 +207,7 @@ let
     (enableFeature enableExecutableProfiling (if versionOlder ghc.version "8" then "executable-profiling" else "profiling"))
     (enableFeature enableSharedLibraries "shared")
     (optionalString (versionAtLeast ghc.version "7.10") (enableFeature doCoverage "coverage"))
-    (optionalString (versionOlder "8.4" ghc.version) (enableFeature enableStaticLibraries "static"))
+    (optionalString (versionOlder "8.4" ghc.version && !stdenv.hostPlatform.isWasm) (enableFeature enableStaticLibraries "static"))
```
[the remark this answers is not in the document]
Comment (dfordivam) [reply pkgs/development/haskell-modules/generic-builder.nix]: Fixed this.

**Comment by Ericson2314** [comment]
@dfordivam Looks good! Could you just update the other GHCs like I mentioned in https://github.com/NixOS/nixpkgs/pull/92168#pullrequestreview-442614430 ? Sorry I took so long to get back to your PRs.

**Comment by dfordivam** [comment]
> 
> 
> @dfordivam Looks good! Could you just update the other GHCs like I mentioned in [#92168 (review)](https://github.com/NixOS/nixpkgs/pull/92168#pullrequestreview-442614430) ? Sorry I took so long to get back to your PRs.

@Ericson2314  I did not change the 8.6.5 as it has the `dontStrip` set to `true`, and my "fix" of the `STRIP_CMD` will effect it.
ref: https://github.com/NixOS/nixpkgs/pull/64371/files

**Comment by FRidh** [comment]
Please rebase. I leave it to the Haskell maintainers to merge it into staging or the haskell-updates branch.

**Comment by Ericson2314** [comment]
Sorry I didn't see the new version earlier!

**Pull request merged**